src/github.com/matrix-org/dendrite/roomserver/input/state.go

```diff
@@ -0,0 +1,258 @@
+			// 3) None of the previous events were state events and they all
+			// have the same state, so this event has exactly the same state
+			// as the previous events.
+			// This should be the common case.
+			return prevState.BeforeStateSnapshotNID, nil
+		}
+		// The previous event was a state event so we need to store a copy
+		// of the previous state updated with that event.
+		stateDataNIDLists, err := db.StateDataNIDs([]types.StateSnapshotNID{prevState.BeforeStateSnapshotNID})
+		if err != nil {
+			return 0, err
+		}
+		stateDataNIDs := stateDataNIDLists[0].StateDataNIDs
+		if len(stateDataNIDs) < maxStateDataNIDs {
+			// 4) The number of state data blocks is small enough that we can just
+			// add the state event as a block of size one to the end of the blocks.
+			return db.AddState(
+				roomNID, stateDataNIDs, []types.StateEntry{prevState.StateEntry},
+			)
+		}
+		// If there are too many deltas then we need to calculate the full state
+		// So fall through to calculateAndStoreStateMany
+	}
+	return calculateAndStoreStateMany(db, roomNID, prevStates)
+}
+
+const maxStateDataNIDs = 64
+
+func calculateAndStoreStateMany(db RoomEventDatabase, roomNID types.RoomNID, prevStates []types.StateAtEvent) (types.StateSnapshotNID, error) {
+	// Conflict resolution.
+	// First stage: load the state datablocks for the prev events.
+	stateNIDs := make([]types.StateSnapshotNID, len(prevStates))
+	for i, state := range prevStates {
+		stateNIDs[i] = state.BeforeStateSnapshotNID
+	}
+	stateDataNIDLists, err := db.StateDataNIDs(uniqueStateSnapshotNIDs(stateNIDs))
+	if err != nil {
+		return 0, err
+	}
+
+	var stateDataNIDs []types.StateDataNID
+	for _, list := range stateDataNIDLists {
+		stateDataNIDs = append(stateDataNIDs, list.StateDataNIDs...)
+	}
+	stateEntryLists, err := db.StateEntries(uniqueStateDataNIDs(stateDataNIDs))
+	if err != nil {
+		return 0, err
+	}
+	stateDataNIDsMap := stateDataNIDListMap(stateDataNIDLists)
+	stateEntriesMap := stateEntryListMap(stateEntryLists)
+
+	var combined []types.StateEntry
+	for _, prevState := range prevStates {
+		list, ok := stateDataNIDsMap.lookup(prevState.BeforeStateSnapshotNID)
+		if !ok {
+			// This should only get hit if the database is corrupt.
+			// It should be impossible for an event to reference a NID that doesn't exist
+			panic(fmt.Errorf("Corrupt DB: Missing state numeric ID %d", prevState.BeforeStateSnapshotNID))
+		}
+
+		var fullState []types.StateEntry
+		for _, stateDataNID := range list {
+			entries, ok := stateEntriesMap.lookup(stateDataNID)
+			if !ok {
+				// This should only get hit if the database is corrupt.
+				// It should be impossible for an event to reference a NID that doesn't exist
+				panic(fmt.Errorf("Corrupt DB: Missing state numeric ID %d", prevState.BeforeStateSnapshotNID))
+			}
+			fullState = append(fullState, entries...)
+		}
+		if prevState.EventStateKeyNID != 0 {
+			fullState = append(fullState, prevState.StateEntry)
```

Comment:
Come on, explain why you're doing this: you want the full state *after* the event in `prev_events` but *before* the event being added in this function, amirite? It's really unclear that `prevState.EventStateKeyNID != 0` really means `prevState.IsStateEvent()` - A tiny method on `types.StateEntry` would go far in improving clarity to prevent me from nagging you to comment every time you do this otherwise.